File: benchmarks/comparison/cpp/02_sieve.cpp
// Sieve of Eratosthenes Benchmark
// Same algorithm as Seen: int64_t flag array, limit=10000000
#include <cstdio>
#include <cstdint>
#include <vector>
#include <chrono>

static std::vector<int64_t> sieve_of_eratosthenes(int64_t limit) {
    std::vector<int64_t> is_prime((size_t)(limit + 1), 1);
    is_prime[0] = 0;
    is_prime[1] = 0;

    for (int64_t p = 2; p * p <= limit; p++) {
        if (is_prime[(size_t)p] != 0) {
            for (int64_t j = p * p; j <= limit; j += p) {
                is_prime[(size_t)j] = 0;
            }
        }
    }

    std::vector<int64_t> primes;
    for (int64_t idx = 2; idx <= limit; idx++) {
        if (is_prime[(size_t)idx] != 0) {
            primes.push_back(idx);
        }
    }
    return primes;
}

static int64_t compute_checksum(const std::vector<int64_t>& primes) {
    int64_t sum = 0;
    for (auto p : primes) sum += p;
    return sum;
}

int main() {
    int64_t limit = 10000000;

    printf("Sieve of Eratosthenes Benchmark\n");
    printf("Finding primes up to: %ld\n", (long)limit);

    int warmup_runs = 3;
    printf("Warming up (%d runs)...\n", warmup_runs);
    for (int w = 0; w < warmup_runs; w++) {
        auto p = sieve_of_eratosthenes(limit);
    }

    printf("Running measured iterations...\n");
    int iterations = 5;
    double min_time = 1e18;
    std::vector<int64_t> result_primes;

    for (int i = 0; i < iterations; i++) {
        auto start = std::chrono::high_resolution_clock::now();
        auto primes = sieve_of_eratosthenes(limit);
        auto end = std::chrono::high_resolution_clock::now();
        double elapsed = std::chrono::duration<double, std::milli>(end - start).count();

        if (elapsed < min_time) {
            min_time = elapsed;
            result_primes = std::move(primes);
        }
    }

    int64_t prime_count = (int64_t)result_primes.size();
    int64_t checksum = compute_checksum(result_primes);

    printf("Prime count: %ld\n", (long)prime_count);
    printf("Checksum: %ld\n", (long)checksum);
    printf("Min time: %.6f ms\n", min_time);
    return 0;
}

File: benchmarks/comparison/zig/02_sieve.zig
// Sieve of Eratosthenes Benchmark
// Same algorithm as C/Seen: bool flag array, limit=10,000,000
const std = @import("std");

fn sieveOfEratosthenes(allocator: std.mem.Allocator, limit: usize) ![]i64 {
    const is_prime = try allocator.alloc(bool, limit + 1);
    defer allocator.free(is_prime);
    @memset(is_prime, true);
    is_prime[0] = false;
    is_prime[1] = false;

    var p: usize = 2;
    while (p * p <= limit) : (p += 1) {
        if (is_prime[p]) {
            var j: usize = p * p;
            while (j <= limit) : (j += p) {
                is_prime[j] = false;
            }
        }
    }

    // Count primes first
    var count: usize = 0;
    for (2..limit + 1) |idx| {
        if (is_prime[idx]) count += 1;
    }

    // Allocate and fill
    const primes = try allocator.alloc(i64, count);
    var pi: usize = 0;
    for (2..limit + 1) |idx| {
        if (is_prime[idx]) {
            primes[pi] = @intCast(idx);
            pi += 1;
        }
    }

    return primes;
}

fn computeChecksum(primes: []const i64) i64 {
    var sum: i64 = 0;
    for (primes) |p| {
        sum += p;
    }
    return sum;
}

pub fn main() !void {
    const stdout = std.fs.File.stdout().deprecatedWriter();
    const allocator = std.heap.page_allocator;

    const limit: usize = 10_000_000;

    try stdout.print("Sieve of Eratosthenes Benchmark\n", .{});
    try stdout.print("Finding primes up to: {d}\n", .{limit});

    // Warmup
    const warmup_runs = 3;
    try stdout.print("Warming up ({d} runs)...\n", .{warmup_runs});
    for (0..warmup_runs) |_| {
        const p = try sieveOfEratosthenes(allocator, limit);
        allocator.free(p);
    }

    // Measured iterations
    try stdout.print("Running measured iterations...\n", .{});
    const iterations = 3;
    var min_time: f64 = 1e18;
    var result_primes: ?[]i64 = null;

    for (0..iterations) |_| {
        var timer = try std.time.Timer.start();
        const primes = try sieveOfEratosthenes(allocator, limit);
        const elapsed_ns = timer.read();
        const elapsed_ms = @as(f64, @floatFromInt(elapsed_ns)) / 1_000_000.0;

        if (elapsed_ms < min_time) {
            min_time = elapsed_ms;
            if (result_primes) |rp| allocator.free(rp);
            result_primes = primes;
        } else {
            allocator.free(primes);
        }
    }

    const primes = result_primes.?;
    const prime_count = primes.len;
    const checksum = computeChecksum(primes);

    try stdout.print("Primes found: {d}\n", .{prime_count});
    try stdout.print("Checksum: {d}\n", .{checksum});
    try stdout.print("Min time: {d:.6} ms\n", .{min_time});

    allocator.free(primes);
}
